internal/state/v1/state.go

```diff
@@ -271,6 +272,74 @@ func (t *tx) TokenByIndex(contract common.Address, idx int) (*big.Int, error) {
 	return enumeratedTotalTree.TokenByIndex(idx)
 }
 
+// TagRoot tags roots for all tree merkle trees at the same block
+func (t *tx) TagRoot(contract common.Address, blockNumber int64) error {
+	slog.Debug("TagRoot ", "contract", contract.String(), "blockNumber", strconv.FormatInt(blockNumber, 10))
+	enumeratedTree, ok := t.enumeratedTrees[contract]
+	if !ok {
+		return fmt.Errorf("contract %s does not exist", contract.String())
+	}
+
+	err := enumeratedTree.TagRoot(blockNumber)
+	if err != nil {
+		return err
+	}
+
+	enumeratedTotalTree, ok := t.enumeratedTotalTrees[contract]
+	if !ok {
+		return fmt.Errorf("contract %s does not exist", contract.String())
+	}
+
+	err = enumeratedTotalTree.TagRoot(blockNumber)
+	if err != nil {
+		return err
+	}
+
+	ownershipTree, ok := t.ownershipTrees[contract]
+	if !ok {
+		return fmt.Errorf("contract %s does not exist", contract.String())
+	}
+
+	return ownershipTree.TagRoot(blockNumber)
+}
+
+// Checkout set current roots the one that are tagged for the block
+// if tag for the block does not exist it searches the first block in the past that has tag
```

Comment:
Can you ask Chat GPT to review the grammar and rephrase it a bit, please? It is a bit confusing for first-time readers in my opinion

Reply:
changed